### ISSUE TITLE
REST (and possibly web) generator doesn't handle functions overload properly

### ISSUE THREAD

**Issue by Geod24** [rejectedsoftware/vibe.d#917]
Pinging @Dicebot 

```
import vibe.d;

@rootPathFromName
interface ITestAPI {
        string getBug();
        string getBug(string par);
        string getBug(int par);
}

class TestAPI : ITestAPI {
override:
        string getBug() { return "foo"; }
        string getBug(string par) { return "bar"; }
        string getBug(int par) { return "foobar"; }
}

shared static this()
{
        auto settings = new HTTPServerSettings;
        settings.port = 8080;
        settings.bindAddresses = ["::1", "127.0.0.1"];

        setLogLevel(LogLevel.debug_);
        auto routes = new URLRouter();
        registerRestInterface(routes, new TestAPI());

        listenHTTP(settings, routes);
        setTimer(1.seconds, () {
                        auto api = new RestInterfaceClient!ITestAPI("http://127.0.0.1:8080");
                        assert("foo" == api.getBug());
                        assert("bar" == api.getBug("Hi"));
                        assert("foobar" == api.getBug(42));
        });
```

Results in:

```
Match tree has 17 nodes, 3 terminals
REST call: GET http://127.0.0.1:8080/i_test_api/bug -> 200, "foo"
REST call: GET http://127.0.0.1:8080/i_test_api/bug?par=Hi -> 200, "foo"
Task terminated with unhandled exception: Assertion failure
Full error: core.exception.AssertError@source/app.d(42): Assertion failure
```

That's because overload will only work in the following conditions:
- Overloads don't resolve to the same templated path (i.e., when there are different path parameters);
- Overloads don't use the same method (i.e. with `@property`).

Will work on a solution.

**Comment by s-ludwig** [comment]
I don't have anything specific in mind, there are a few old issues that are not particularly difficult, such as #266 or #876, but the more recent ones are usually simpler to solve. So apart from those two, I'd rather look only at the issues from this year.

**Comment by s-ludwig** [comment]
@Geod24  w.r.t. `URLRouter` asserting that a route is only registered once; the router is especially designed so that routes can fall through to the next matching route, regardless if the route string is the same or different. This enables some very useful use cases.

BTW, since the data comes in as a string, overloads that are generally not ambiguous may very well be in the context of the REST interface. How would you know for example which of two overloads is the right one both accept two different UDTs with `fromString` methods. Checking for exceptions during the string conversion would be an option, but would also yield to completely different overloading rules than for normal D code.

Currently I tend to think that any possibly ambiguous overloads should be forbidden - in the simplest conservative approach, overloads could simply be enforced to always differ in their `@path` or `@method`. This would be a bit less convenient, but at least we avoid obscure behavior... these interface generators are already quite complex after all

**Comment by MartinNowak** [comment]
> Currently I tend to think that any possibly ambiguous overloads should be forbidden - in the simplest conservative approach, overloads could simply be enforced to always differ in their @path or @method.

I agree, overloading by query params is a bad idea, because we're not dealing with typed data here.
